### DISCUSSION TITLE
Make achievement sets per hash, rather than per game, for easier multi-region support

### DISCUSSION THREAD

**Discussion by kingspoons** [RetroAchievements/RAIntegration#966] [Ideas]
Hi all,

I'm not 100% sure which branch of retroachievements my suggestion should be related to - hopefully if it's in the wrong section it can be moved.

**What**

After speaking with a few other developers, I'd like to pitch the idea of changing achievement sets to be related to games on a **per hash** basis, rather than **per game** as it is now. 

**Why**

A while back, I joined the compatibility testing team, with the hope of being able to check EU (or UK) versions of roms that I own against the sets already on the site. As most games use the US hash (and I'd just made myself a [Sanni cart reader](https://github.com/sanni/cartreader) to dump my cart collection), I was actually really excited about getting stuck in, hoping to not only have some fun, but that I'd also help the community in general. With official threads on the forums often being [full](https://retroachievements.org/viewtopic.php?t=185&o=60) of [requests](https://retroachievements.org/viewtopic.php?t=4753&c=137173#137173) asking to [link](https://retroachievements.org/viewtopic.php?t=4701&c=56310#56310) hashes in specific ([mostly EU](https://retroachievements.org/viewtopic.php?t=4701&c=63720#63720)) [languages](https://retroachievements.org/viewtopic.php?t=11357), it's clear "more (regional) hashes!" was something the community were really keen for.

However, I soon hit a few problems.

- Sometimes, the game would be an exact match, but the devs either didn't reply to my messages, or were no longer on the site. (This was the case for Super Metroid, where the EU rom appears to be an exact match. Adding the hash would add German and French localisations, so it's really cool low hanging fruit, but I've had no reply from the devs.)
- Sometimes, the game was a close match, and would only require a few addresses to be changed to make it compatible (e.g. Battle for Naboo on N64, Star Wars: Episode 1 on PS1). Sometimes the devs were more than happy to help; other times, they outright refused, as they didn't want to open themselves up to any unseen issues if it turns out later that things worked slightly differently on the EU rom, and they ended up with strikes against their name. 
- Sometimes, the memory was completely different - but even in this scenario, it's a thousand times faster to find those same addresses in the EU rom than it is to do the hard work hunting them down to begin with.

So bar a handful of games, pretty much every game I tried to add ended up hitting one of these roadblocks. But it's something that I think could be easily(?) fixed by changing sets to be linked on a per hash, rather than a per game basis.

That way, any identical versions could easily be added without the original dev having to take any responsibility, while EU versions that need changes could basically be a fork of the original, tweaking the memory locations it needs to tweak while keeping the rest intact. Whether the new hashes are entirely separate, and we just copy and paste the achievements over, changing what needs to be changed as we go, or we try something more complex like a diff list, I don't know, but either way - I am dying to be able to spend some time poring over my collection and making them compatible with the sets on the site, adding multi-language (and region) support to many a game, but it looks like it'd require a bit of a tweak to the system first.

**Benefits**

- Easier to add support for roms with multiple languages (one of the biggest reasons people suggest adding new hashes)
- Including more languages expands the retroachievements audience (and therefore brings in more devs!)
- Makes it far easier for people to use dumps of their own games (especially for PS1/PS2, where the vast majority of hashes used are for US roms, yet it's incredibly easy to dump your own discs), rather than requiring the use of potentially shady websites and/or piracy. 
- Makes roms more likely to "just work", rather than having to hunt down specific versions

It seems like the sort of thing that'd be a really strong addition to the site, and help retroachievements reach even more people around the globe. I'm not sure how much work would be required on the back end to make it a reality, but another dev I was talking to mentioned that a feature like this was planned at one point, but he hadn't heard anything about it since. I just wanted to flag it up here as an idea, in the (misplaced?) hope that it may be a fairly straightforward thing to do!

**Comment by Jamiras** [comment]
Hash management is server-side, so this should have been submitted to the RAWeb repository.

There's already a few discussions there that might be relevant:
* https://github.com/RetroAchievements/RAWeb/discussions/742
* https://github.com/RetroAchievements/RAWeb/discussions/789
* https://github.com/RetroAchievements/RAWeb/discussions/717
* https://github.com/RetroAchievements/RAWeb/discussions/781

**Comment by Jamiras** [comment]
Also, per your title "for easier multi-region support" - that just isn't true. If each region has its own set of achievement definitions, you end up with either a duplicate set that can earn the player "double points", or have to have some way to share the unlocks across the definitions. Additionally, if there's multiple definitions of an achievement, a bugfix made in one is less likely to be considered for the others.

**Reply by kingspoons** [reply]
I don't really know how the logic works behind the scenes, so it's hard to try and come up with a solution that would work without knowing how the pieces fit together. That being said, I imagine it would be possible to have the game ID as the main "thing" achievements are linked to, and then have it load up a different (or rather, regional variant) set depending on the version of the game you're playing.

For the second point, I agree, and I think that's the biggest issue, but there ought to be a way around it. Could be as easy as just notifying developers about any issues that come up with other branches of the game.
